### ISSUE TITLE
Parallelizing NiftiSpheres/LabelsMasker?

### ISSUE THREAD

**Issue by dPys** [nilearn/nilearn#1501]
Is there any way to speed up NiftiSpheresMasker and NiftiLabelsMasker (e.g. via joblib, iterables, etc.)? Maybe this functionality already exists or is in the works, but I haven't been able to find it.  If it does not yet exist, this would be a very cool feature to incorporate. I would think that these functions would be mpi-parallelizable across coordinates (i.e. the total runtime should theoretically be reducible to the length of time that it takes to extract signal from a single sphere/label assuming that number of coordinates = # cores?)

**Comment by KamalakerDadi** [comment]
- Yes, PR will come soon. All masker families should benefit via joblib caching and parallelization and multi-subject extraction. This is part of issue opened at #1240.

**Comment by bthirion** [comment]
- Addressed in #2294.

**Comment by bthirion** [comment]
'n_jobs' argument could be added to NiftiSpheresMasker though.

**Comment by dPys** [comment]
@bthirion -- were you thinking [here](https://github.com/nilearn/nilearn/blob/55e3f26dbd9fc6e89516e5f37e8aae23ec6086a9/nilearn/regions/signal_extraction.py#L124-L127)

**Comment by bthirion** [comment]
Naively I was thinking about iterating across spheres in the SpheresMasker but this leads to bad memory usage, so indeed, you suggestion might be better.

**Comment by dPys** [comment]
#2693

**Comment by Remi-Gau** [comment]
I wonder if it may be good to revisit this one during the sprint: for images with a lot of images or labels, we may get some speed up from parallelizing over image / label.

The problem is that this will impact MultiMaskers that are parallelizing over subjects: so when user are trying to leverage n_jobs for Multimaskers is the n_jobs supposed to parallelize over subjects or labels? do we hard code that decision or do we let the user choose dependending on where their bottle neck is?

I think this could potentially be a low-hanging fruit where we can get some performance gain.

**Comment by bthirion** [comment]
But NiftiSpheres and NiftiLablesMasker (actually all non-Multi-maskers) are not designed to deal with multi-subjects settings primarily. So, it would make sense to parallelize on regions/spheres/labels instead for those.

**Comment by Remi-Gau** [comment]
> But NiftiSpheres and NiftiLablesMasker (actually all non-Multi-maskers) are not designed to deal with multi-subjects settings primarily. So, it would make sense to parallelize on regions/spheres/labels instead for those.

yes yes I agree and I think we should do it but because the multi maskers inherit from the non-multi counter part, if we just use an 'n_jobs' parameters / attributes we will also affect the multi maskers so we may have to plan this a bit
